### PR TITLE
LowProFool, imperceptible attack on tabular data

### PR DESCRIPTION
**Pull request by marcinz99**
# Description

New evasion attack is added - LowProFool, imperceptible attack on tabular data.
>`LowProFool` attack transforms the provided real-valued tabular data into adversaries of the specified target classes. The generated adversaries have to be as close as possible to the original samples in terms of the weighted Lp-norm, where the weights determine each feature's importance.
>
>| Paper link: https://arxiv.org/abs/1911.03274

PR includes:
* attack's implementation,
* unit tests,
* notebook presenting usage of the new feature.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Tests employing diffentent underlying classifiers and datasets.

**Test Configuration**:
- OS - Windows 10
- Python version - 3.7.10
- ART version or commit number - 1.6.1
- PyTorch version - 1.3.1

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
Hi @marcinz99 Thank you very much for you pull request, we'll start reviewing it as soon as possible.

File: tests/attacks/evasion/test_lowprofool.py [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+            logger.info("[Iris flower, Logistic Regression (custom)] Clipping is invalid.")
+
+        self.assertTrue(is_valid_1, msg="Bad clipping")
+        self.assertTrue(is_valid_2, msg="Bad clipping")
+        self.assertTrue(is_valid_3, msg="Bad clipping")
+        self.assertTrue(is_valid_4, msg="Bad clipping")
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:585]:
```suggestion
```

File: tests/attacks/evasion/test_lowprofool.py [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+            "{:.2f}%".format(expected*100, success_rate * 100)
+        )
+
+        print("===== T E S T =====")
+        print("Test name:           {}".format(self.id()))
+        print("Classifier:          PyTorch neural network")
+        print("Dataset:             Iris flower")
+        print("Validation set size: {}".format(target.shape[0]))
+        print("Success rate:        {:.2f}%".format(100 * success_rate))
+        print()
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:377]: Please replace `print` with `logger.info`.

File: tests/attacks/evasion/test_lowprofool.py [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+            "{:.2f}%".format(expected*100, success_rate * 100)
+        )
+
+        print("===== T E S T =====")
+        print("Test name:           {}".format(self.id()))
+        print("Classifier:          Logistic Regression (custom)")
+        print("Dataset:             Breast cancer")
+        print("Validation set size: {}".format(target.shape[0]))
+        print("Success rate:        {:.2f}%".format(100 * success_rate))
+        print()
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:325]: Please replace `print` with `logger.info`.

File: tests/attacks/evasion/test_lowprofool.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+        logger.info("[Wines, Logistic Regression (custom)] success rate of adversarial attack (expected >{:.2f}):"
+                    " {:.2f}%".format(expected*100, success_rate * 100))
+
+        print("===== T E S T =====")
+        print("Test name:           {}".format(self.id()))
+        print("Classifier:          Logistic Regression (custom)")
+        print("Dataset:             Wines")
+        print("Validation set size: {}".format(target.shape[0]))
+        print("Success rate:        {:.2f}%".format(100 * success_rate))
+        print()
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:280]: Please replace `print` with `logger.info`.

File: tests/attacks/evasion/test_lowprofool.py [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+        logger.info("[Irises, Logistic Regression (custom)] success rate of adversarial attack (expected >{:.2f}): "
+                    "{:.2f}%".format(expected*100, success_rate * 100))
+
+        print("===== T E S T =====")
+        print("Test name:           {}".format(self.id()))
+        print("Classifier:          Logistic Regression (custom)")
+        print("Dataset:             Iris flower")
+        print("Validation set size: {}".format(target.shape[0]))
+        print("Success rate:        {:.2f}%".format(100 * success_rate))
+        print()
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:237]: Please replace `print` with `logger.info`.

File: tests/attacks/evasion/test_lowprofool.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+
+    def predict(self, x: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Perform prediction of the estimator for input `x`.
+
+        :param x: Samples.
+        :param kwargs:
+        :return: Predictions by the model.
+        """
+        return self.model.predict_proba(x)
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:149]: Please replace with `art.estimators.classification.SklearnLogisticRegression` updated with branch `development_issue_1064`.

File: tests/attacks/evasion/test_lowprofool.py [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+            y_pred = nn_model.forward(x)
+            loss = self.loss_fn(y_pred, y)
+            nn_model.zero_grad()
+            loss.backward()
+            with torch.no_grad():
+                for param in nn_model.parameters():
+                    param.data -= learning_rate * param.grad
+
+
+class TestLowProFool(unittest.TestCase):
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:176]: Please convert to pytest format similar to the other tests in `tests/attacks/evasion`.

File: art/attacks/evasion/lowprofool.py [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+import logging
+import numpy as np
+from scipy.stats import pearsonr
+from tqdm.auto import trange
+
+from art.attacks.attack import EvasionAttack
+from art.estimators.estimator import LossGradientsMixin
+from art.estimators.estimator import BaseEstimator
+from art.estimators.classification.classifier import ClassifierMixin
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union, Type, TYPE_CHECKING
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:39]:
```suggestion
import logging
from typing import Any, Callable, Dict, List, Optional, Tuple, Union, Type, TYPE_CHECKING

import numpy as np
from scipy.stats import pearsonr
from tqdm.auto import trange

from art.attacks.attack import EvasionAttack
from art.estimators.estimator import LossGradientsMixin
from art.estimators.estimator import BaseEstimator
from art.estimators.classification.classifier import ClassifierMixin
```

File: art/attacks/evasion/lowprofool.py [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+logger = logging.getLogger(__name__)
+
+
+class LowProFool(EvasionAttack):
+    """
+    `LowProFool` attack.
+
+    | Paper link: https://arxiv.org/abs/1911.03274
+    """
+    attack_params = EvasionAttack.attack_params + [
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:53]:
```suggestion
    """

    attack_params = EvasionAttack.attack_params + [
```

File: art/attacks/evasion/lowprofool.py [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+            classifier: "CLASSIFIER_CLASS_LOSS_GRADIENTS_TYPE",
+            n_steps: Optional[Optional[int]] = 100,
+            threshold: Optional[Optional[Union[float, None]]] = 0.5,
+            lambd: Optional[float] = 1.5,
+            eta: Optional[float] = 0.2,
+            eta_decay: Optional[float] = 0.98,
+            eta_min: Optional[float] = 1e-7,
+            p: Optional[Union[int, float, str]] = 2,
+            importance: Optional[Union[Callable, str, np.ndarray]] = 'pearson',
+            verbose: Optional[bool] = False
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:77]:
```suggestion
        self,
        classifier: "CLASSIFIER_CLASS_LOSS_GRADIENTS_TYPE",
        n_steps: Optional[Optional[int]] = 100,
        threshold: Optional[Optional[Union[float, None]]] = 0.5,
        lambd: Optional[float] = 1.5,
        eta: Optional[float] = 0.2,
        eta_decay: Optional[float] = 0.98,
        eta_min: Optional[float] = 1e-7,
        p: Optional[Union[int, float, str]] = 2,
        importance: Optional[Union[Callable, str, np.ndarray]] = "pearson",
        verbose: Optional[bool] = False,
```

File: art/attacks/evasion/lowprofool.py [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+        return self.lambd * np.linalg.norm(
+            self.importance_vec * perturbations, axis=1,
+            ord=(np.inf if self.p == "inf" else self.p)
+        ).reshape(-1, 1)
+
+    def __classification_loss_gradient(
+            self,
+            samples: np.ndarray,
+            perturbations: np.ndarray,
+            y_probas: np.ndarray,
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:173]: The argument `y_probas` seems unused. Is this ok or can we remove `y_probas`?

File: art/attacks/evasion/lowprofool.py [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+
+        :param samples: Base design matrix.
+        :param perturbations: Perturbations of samples towards being adversarial.
+        :param y_probas: Class-wise prediction probabilities.
+        :param targets: The target labels for the attack.
+        :return: Array of classification loss gradients.
+        """
+        return self.estimator.loss_gradient(
+            (samples + perturbations).astype(np.float32),
+            (targets).astype(np.float32)
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:187]:
```suggestion
            targets.astype(np.float32)
```

File: art/attacks/evasion/lowprofool.py [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+
+        # Predicate used to determine whether the target was met based on the given probabilities
+        def met_target(y_probas, target_class):
+            if self.threshold is None:
+                return np.argmax(y_probas) == target_class
+            else:
+                return y_probas[target_class] > self.threshold
+
+        # Main loop.
+        for i in trange(self.n_steps, desc="LowProFool", disable=not(self.verbose)):
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:407]:
```suggestion
        for i in trange(self.n_steps, desc="LowProFool", disable=not self.verbose):
```

File: art/attacks/evasion/lowprofool.py [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+        This function allows one to easily calculate the feature importance vector using the pre-specified function,
+        in case it wasn't passed at initialization.
+
+        :param x: Design matrix of the dataset used to train the classifier.
+        :param y: Labels of the dataset used to train the classifier.
+        :param importance_array: Array providing features' importance score.
+        :param normalize: Assure that feature importance values sum to 1.
+        :return: LowProFool instance itself.
+        """
+        if not (importance_array is None):
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:336]:
```suggestion
        if importance_array is not None:
```

File: art/attacks/evasion/lowprofool.py [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+    def __init__(
+            self,
+            classifier: "CLASSIFIER_CLASS_LOSS_GRADIENTS_TYPE",
+            n_steps: Optional[Optional[int]] = 100,
+            threshold: Optional[Optional[Union[float, None]]] = 0.5,
+            lambd: Optional[float] = 1.5,
+            eta: Optional[float] = 0.2,
+            eta_decay: Optional[float] = 0.98,
+            eta_min: Optional[float] = 1e-7,
+            p: Optional[Union[int, float, str]] = 2,
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:75]: Please replace `p` with `norm` similar to the other evasion attacks.

File: art/attacks/evasion/lowprofool.py [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+        :param true: Ground truth (correct) labels for samples.
+        :param predicted: Predicted probabilities.
+        :return: Mean cross entropy.
+        """
+        p = np.array([1 - true, true])
+        q = np.array([1 - predicted, predicted])
+
+        return np.mean(
+            -1 * np.sum(p * np.log(q + 1e-13), axis=0), axis=1
+        ).reshape(-1, 1)
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:152]: Could this method or parts of it be replaced with the existing `sklearn.metrics.log_loss`?

File: art/attacks/evasion/lowprofool.py [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+
+        :param perturbations: Perturbations of samples towards being adversarial.
+        :return: Array with weighted Lp-norm of perturbations.
+        """
+        return self.lambd * np.linalg.norm(
+            self.importance_vec * perturbations, axis=1,
+            ord=(np.inf if self.p == "inf" else self.p)
+        ).reshape(-1, 1)
+
+    def __classification_loss_gradient(
```
Comment (beat-buesser) [inline art/attacks/evasion/lowprofool.py:169]: This method is just one line and used only once. Please inline the call to `self.estimator.loss_gradient` and replace the call to `__classification_loss_gradient`.

File: tests/attacks/evasion/test_lowprofool.py [hunk 18]
```diff
@@ -0,0 +1,585 @@
...
+from sklearn.model_selection import StratifiedShuffleSplit
+from sklearn.linear_model import LogisticRegression
+from sklearn.svm import SVC
+from scipy.stats import pearsonr
+from art.attacks.evasion import LowProFool
+
+from art.estimators.estimator import LossGradientsMixin
+from art.estimators.estimator import BaseEstimator
+from art.estimators.classification.classifier import ClassifierMixin
+from art.estimators.classification.scikitlearn import ScikitlearnSVC
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:40]:
```suggestion
import logging
import unittest

import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from torch.autograd import Variable
from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from scipy.stats import pearsonr

from art.attacks.evasion import LowProFool
from art.estimators.classification import PyTorchClassifier
from art.estimators.estimator import LossGradientsMixin
from art.estimators.estimator import BaseEstimator
from art.estimators.classification.classifier import ClassifierMixin
from art.estimators.classification.scikitlearn import ScikitlearnSVC
```

File: tests/attacks/evasion/test_lowprofool.py [hunk 19]
```diff
@@ -0,0 +1,585 @@
...
+
+    def train_nn(self, nn_model, x, y, learning_rate, epochs):
+        for _ in range(epochs):
+            y_pred = nn_model.forward(x)
+            loss = self.loss_fn(y_pred, y)
+            nn_model.zero_grad()
+            loss.backward()
+            with torch.no_grad():
+                for param in nn_model.parameters():
+                    param.data -= learning_rate * param.grad
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_lowprofool.py:173]:
```suggestion
    optimizer = optim.SGD(nn_model.parameters(), lr=0.001, momentum=0.9)
    
    for _ in range(epoches):
        y_pred = nn_model.forward(X)
        loss = self.loss_fn(y_pred, y)
        nn_model.zero_grad()
        loss.backward()
        
        optimizer.step()
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @marcinz99 Thank you very much for implementing the `LowProFool` evasion attack, the first attack specifically for tabular data! Your code is well organised and follows the patterns of ART very well.

I have made a few change request inline above. We should update this PR with the upcoming changes to ART 1.6.2 on branch `development_issue_1064` to resolve #1064 opened by you recently.

Following the updates for #1064, could you please remove cell 18 of the notebook and the markdown comment cell before cell 18?

Could you also please take a look at the failing style check in the CI runs and address the warnings raised by `mypy`, `pylint` and `pycodestyle`? Please run first Black formatter `black==20.8b1` on `art/` and `tests/` with command `black --line-length 120 path` which should address some of the warnings.

**Comment by marcinz99** [comment]
Hi @beat-buesser Thanks for the comprehensive feedback. We will try to go through your comments and remarks shortly, but I have to inform you that as for the time being we have lots of work related to the current university courses. Therefore we will apply the necessary corrections to the code no sooner than the next week or so.

Regarding the cell 18 of the notebook (custom scikit-learn logistic regression wrapper), we thought that it would be nice to 
actually show the user how to create a custom wrapper, just in case they couldn't use any built-in one. Especially as we had priorly created one ourselves and it turned out to be pretty concise. Do you find it reasonable to leave it there to serve such a purpose then?

**Comment by beat-buesser** [comment]
> Regarding the cell 18 of the notebook (custom scikit-learn logistic regression wrapper), we thought that it would be nice to
> actually show the user how to create a custom wrapper, just in case they couldn't use any built-in one. Especially as we had priorly created one ourselves and it turned out to be pretty concise. Do you find it reasonable to leave it there to serve such a purpose then?

Hi @marcinz99 I think it is better to remove it, because it us replicating an existing estimator of ART and because the focus of the notebook should be on `LowProFool`.

**Comment by beat-buesser** [comment]
@marcinz99 This pull request contains a new tool, therefore you can add the name of your university to the [AUTHORS](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/AUTHORS) file if you like.

**Review by beat-buesser — APPROVED** [review]
Hi @marcinz99, @jupiterb, @czyjtu, @kocimski, and @ArcziPT

Congratulations for your first contribution to ART and thank you very much for implementing the `LowProFool` evasion attack on tabular data and addressing all review comments! Great work!

**Pull request merged**